opt_einsum/paths.py

```diff
@@ -272,3 +279,199 @@ def greedy(input_sets, output_set, idx_dict, memory_limit):
+    remaining = {}  # key -> ssa_id
+    ssa_ids = itertools.count(len(inputs))
+    ssa_path = []
+    for ssa_id, key in enumerate(inputs):
+        if key in remaining:
+            ssa_path.append((remaining[key], ssa_id))
+            remaining[key] = next(ssa_ids)
+        else:
+            remaining[key] = ssa_id
+
+    # Compute footprints of each tensor.
+    footprints = {key: helpers.compute_size_by_dict(key, sizes)
+                  for key in remaining}
+
+    # Keep track of possible contraction dims.
+    dim_to_keys = defaultdict(set)
+    for key in remaining:
+        for dim in key - output:
+            dim_to_keys[dim].add(key)
+
+    # Keep track of the number of tensors using each dim; when the dim is no longer
+    # used it can be contracted. Since we specialize to binary ops, we only care about
+    # ref counts of >=2 or >=3.
+    dim_ref_counts = {
+        count: set(dim for dim, keys in dim_to_keys.items() if len(keys) >= count) - output
+        for count in [2, 3]}
+
+    # Find initial candidate contractions.
+    queue = []
+    for dim, keys in dim_to_keys.items():
+        keys = list(keys)
+        for i, k1 in enumerate(keys):
+            k2s = keys[:i]
+            _push_candidate(output, sizes, remaining, footprints, dim_ref_counts, k1, k2s, queue)
+
+    # Greedily contract pairs of tensors.
+    while queue:
+        cost, k1, k2, k12 = heapq.heappop(queue)
+        if k1 not in remaining or k2 not in remaining:
+            continue  # candidate is obsolete
+
+        ssa_id1 = remaining.pop(k1)
+        ssa_id2 = remaining.pop(k2)
+        for dim in k1 - output:
+            dim_to_keys[dim].remove(k1)
+        for dim in k2 - output:
+            dim_to_keys[dim].remove(k2)
+        ssa_path.append((ssa_id2, ssa_id1))
+        if k12 in remaining:
+            ssa_path.append((remaining[k12], next(ssa_ids)))
+        else:
+            footprints[k12] = helpers.compute_size_by_dict(k12, sizes)
+            for dim in k12 - output:
+                dim_to_keys[dim].add(k12)
+        remaining[k12] = next(ssa_ids)
+        _update_ref_counts(dim_to_keys, dim_ref_counts, k1 | k2 - output)
+
+        # Find new candidate contractions.
+        k1 = k12
+        k2s = set(k2 for dim in k1 for k2 in dim_to_keys[dim] if k2 != k1)
+        _push_candidate(output, sizes, remaining, footprints, dim_ref_counts, k1, k2s, queue)
+
+    # Greedily compute pairwise outer products.
+    queue = [(len(key & output), ssa_id, key) for key, ssa_id in remaining.items()]
+    heapq.heapify(queue)
+    _, ssa_id1, k1 = heapq.heappop(queue)
+    while queue:
+        _, ssa_id2, k2 = heapq.heappop(queue)
+        ssa_path.append((ssa_id1, ssa_id2))
+        k12 = (k1 | k2) & output
+        cost = len(k12)
+        ssa_id12 = next(ssa_ids)
+        _, ssa_id1, k1 = heapq.heappushpop(queue, (cost, ssa_id12, k12))
+
+    # Perform one final reduction to match output shape.
+    ssa_path.append((ssa_id1,))
```

Comment:
Yes see comment above about output indices - I think this shouldn't be necessary.